### PR TITLE
SHiP finality_data: use string encoding of public keys in pending finalizer policy

### PR DESCRIPTION
**Pull request by linh2931**
For simpler and easier uses by SHiP clients, use string encoded public keys of pending finalizer policy in finality_data. It is decided both on-disk SHiP logs and websocket streaming use the same string format, for keeping consistent with other data in SHiP (disk and wire formats are the same), and for simpler implementation.

`ship_protocol.hpp` in abieos submodule will be updated after this PR is merged.

Resolves https://github.com/AntelopeIO/spring/issues/342.

### CODE REVIEW

File: libraries/chain/include/eosio/chain/finality/finalizer_authority.hpp [hunk 1]
```diff
@@ -14,6 +14,19 @@ namespace eosio::chain {
...
    };
 
+   // This is used by SHiP and Deepmind which require public keys in string format.
+   struct finalizer_authority_with_string_key {
+      std::string  description;
+      uint64_t     weight = 0;
+      std::string  public_key;
+
+      finalizer_authority_with_string_key() = default;
+      finalizer_authority_with_string_key(const finalizer_authority& input)
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/finality/finalizer_authority.hpp:24]: Make `explicit`

File: libraries/chain/include/eosio/chain/finality/finalizer_policy.hpp [hunk 2]
```diff
@@ -50,6 +50,27 @@ namespace eosio::chain {
...
    };
 
+   // This is used by SHiP and Deepmind which require public keys in string format.
+   struct finalizer_policy_with_string_key {
+      uint32_t                                         generation = 0;
+      uint64_t                                         threshold = 0;
+      std::vector<finalizer_authority_with_string_key> finalizers;
+
+      finalizer_policy_with_string_key() = default;
+      finalizer_policy_with_string_key(const finalizer_policy& input)
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/finality/finalizer_policy.hpp:60]: Make `excplicit`

**Review by heifner — CHANGES_REQUESTED** [review]

**Review by spoonincode — APPROVED** [review]

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/include/eosio/chain/finality/finalizer_authority.hpp (hunk 1 above)
Comment (linh2931) [reply libraries/chain/include/eosio/chain/finality/finalizer_authority.hpp]: https://github.com/AntelopeIO/spring/pull/348/commits/64b30a41ae1c103c367d648cb5b0073573d6a222

File: libraries/chain/include/eosio/chain/finality/finalizer_policy.hpp (hunk 2 above)
Comment (linh2931) [reply libraries/chain/include/eosio/chain/finality/finalizer_policy.hpp]: https://github.com/AntelopeIO/spring/pull/348/commits/64b30a41ae1c103c367d648cb5b0073573d6a222

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: To support ease of use SHiP, finality_data captures a new pending finality policy, including encoded public keys from pending finality data.
Note:end

**Pull request merged**